File: backend/app/core/calendar.py
"""تولید لینک «افزودن به Google Calendar» — بدون OAuth/API call/ذخیره‌سازی
(تصمیم صریح کاربر، بخش ۸ پلن معماری)."""

from datetime import datetime, timedelta
from urllib.parse import urlencode

_DATE_FORMAT = "%Y%m%dT%H%M%SZ"


def google_calendar_link(
    *, title: str, description: str, location: str, starts_at: datetime, duration_minutes: int
) -> str:
    ends_at = starts_at + timedelta(minutes=duration_minutes)
    params = {
        "action": "TEMPLATE",
        "text": title,
        "dates": f"{starts_at.strftime(_DATE_FORMAT)}/{ends_at.strftime(_DATE_FORMAT)}",
        "details": description,
        "location": location,
    }
    return "https://calendar.google.com/calendar/render?" + urlencode(params)

File: frontend/src/lib/calendar.ts
// معادل فرانتی app/core/calendar.py — تابع محض بدون OAuth/API call، پس نیازی
// به لاگین یا ثبت‌نام برای «افزودن به تقویم» روی صفحه‌ی عمومی رویداد نیست
// (برخلاف افزودن به تقویم بعد از خرید بلیط که چون به یک registration خاص
// وصله، نیاز به لاگین داره — این همون منطق رو برای بازدیدکننده‌ی مهمان تکرار می‌کنه).
function toGoogleDate(iso: string): string {
  return new Date(iso).toISOString().replace(/[-:]/g, "").replace(/\.\d{3}/, "");
}

export function buildGoogleCalendarLink({
  title,
  description,
  location,
  startsAtIso,
  durationMinutes,
}: {
  title: string;
  description: string;
  location: string;
  startsAtIso: string;
  durationMinutes: number;
}): string {
  const start = new Date(startsAtIso);
  const end = new Date(start.getTime() + durationMinutes * 60 * 1000);
  const params = new URLSearchParams({
    action: "TEMPLATE",
    text: title,
    dates: `${toGoogleDate(start.toISOString())}/${toGoogleDate(end.toISOString())}`,
    details: description,
    location,
  });
  return `https://calendar.google.com/calendar/render?${params.toString()}`;
}
